alacritty/src/display/mod.rs

```diff
@@ -1529,6 +1550,54 @@ impl<T> DerefMut for Replaceable<T> {
     }
 }
 
+/// The frame timer state.
+#[derive(Default)]
+pub struct FrameTimerInfo {
+    /// The optimal delay for the renderer.
+    swap_delay: Option<Duration>,
+
+    /// The time stamp of the previous contiguos frame.
+    frame_timestamp: Option<Instant>,
+
+    /// The refresh rate we've used to compute the `swap_delay`.
+    refresh_interval: Duration,
+}
+
+impl FrameTimerInfo {
+    /// Compute the delay that we should use to achieve the target frame
+    /// rate.
+    pub fn compute_delay(&mut self, refresh_interval: Duration) -> Duration {
+        // If the refresh rate has changed, reset the timings.
+        if self.refresh_interval != refresh_interval || self.swap_delay.is_none() {
+            self.refresh_interval = refresh_interval;
+            self.frame_timestamp = Some(Instant::now());
+            self.swap_delay = Some(refresh_interval);
+            return refresh_interval;
+        }
+
+        // At this point we know that `swap_delay` has some value.
+        let swap_delay = self.swap_delay.unwrap();
+
+        let new_swap_delay = match self.frame_timestamp {
+            Some(last_frame_timestamp) => {
+                // Compute the error.
+                let error = self.refresh_interval.as_micros() as i64
+                    - last_frame_timestamp.elapsed().as_micros() as i64;
+                let new_swap_delay = swap_delay.as_micros() as i64 + error;
```

Comment:
Assuming that a frame took a particularly long time, that will mean that immediately the next frame will get hurt a lot too. But I don't think there's any immediate connection between the last frame and the next frame, right?

So shouldn't we try to "stabilize" towards some rolling average rather than flip-flopping between a massive and tiny delay?

What you seem to be doing is basically just to try and **start** rendering every *refresh_rate* ms, while realistically what I think we'd want is to try and **finish** rendering every *refresh_rate* ms.

So I feel like we should add a rolling average to this and subtract it as a "buffer" for how long we'd expect rendering to take. Then keep your logic to adjust for when that buffer was insufficient or to much to ensure we're staying on track. But that way we should be able to aim to finish in time rather than starting in time.

Reply:
>Assuming that a frame took a particularly long time, that will mean that immediately the next frame will get hurt a lot too.

The frame will fire immediately in this case. Not sure how correct this will be though.

>But I don't think there's any immediate connection between the last frame and the next frame, right?

The connection was initially made for contiguous frames, but it was showing worse perf then what it's right now. The initial idea was to compute the optimal `timeout` for contiguous frames and use it as a `default` when we don't know sufficient information about 'how much time we should timeout to render at the monitor refresh rate'.

>So shouldn't we try to "stabilize" towards some rolling average rather than flip-flopping between a massive and tiny delay?

It was slower, but maybe I was doing it wrong. Basically what I was trying to do is when we hit `draw` and we skip it due to frame not being dirty I was resetting the last timestamp. But I think what we could try to do is when the error gets too big as in it zeros the delay we might want to ignore adjusting the delay in this case. The large error means that we likely wasn't ready to draw.

>What you seem to be doing is basically just to try and start rendering every refresh_rate ms, while realistically what I think we'd want is to try and finish rendering every refresh_rate ms.

I think the latter is what I'm trying to do since I schedule the frame after all the rendering, so I'm accounting not for the start but for the end. The place is after all the rendering operations and I'm scheduling for that. If I were doing the same thing before the `swap_buffers` then yes, it'll be the former.

>So I feel like we should add a rolling average to this and subtract it as a "buffer" for how long we'd expect rendering to take. 

Rendering is already accounted since we do that after `gl::Finish` and `swap_buffers`, it could be wise to call the `gl::Finish` on other platforms though, however it should be the case on other platforms already with the swap buffers.


>Then keep your logic to adjust for when that buffer was insufficient or to much to ensure we're staying on track. But that way we should be able to aim to finish in time rather than starting in time.

We aiming on finish time already

So to sum it up, what could be adjusted is behavior when swap delay get zeroed, since it indeed breaks some scheduling, however when I was trying to do so with the method I've described above I got lower fps then with the current approach, which makes sense, since I was rendering with 0 delay the next frame as well or something like that.

I also used various techniques, like using `Meter` for averaging the error, but not sure whether it got me anywhere, I think I should retry at least, since last time I did that with wrong scheduler logic and haven't bothered to retest once I changed the scheduler timing re-computation.